toolchain/check/handle_interface.cpp

```diff
@@ -126,14 +128,34 @@ auto HandleInterfaceDefinitionStart(
   // Enter the interface scope.
   context.scope_stack().Push(interface_decl_id, interface_info.scope_id);
 
-  // TODO: Introduce `Self`.
-
   context.inst_block_stack().Push();
   context.node_stack().Push(parse_node, interface_id);
 
   // We use the arg stack to build the witness table type.
   context.args_type_info_stack().Push();
 
+  // Declare and introduce `Self`.
+  if (!interface_info.is_defined()) {
+    // TODO: Once we support parameterized interfaces, this won't be the right
+    // type. For `interface X(T:! type)`, the type of `Self` is `X(T)`, whereas
+    // this will be simply `X`.
+    auto self_type_id = context.GetTypeIdForTypeConstant(
+        context.constant_values().Get(interface_decl_id));
+
+    // We model `Self` as a symbolic binding whose type is the interface.
+    interface_info.self_param_id = context.AddInst(
+        {Parse::NodeId::Invalid,
+         SemIR::BindSymbolicName{
+             self_type_id,
+             context.bind_names().Add(
+                 {.name_id = SemIR::NameId::SelfType,
+                  .enclosing_scope_id = interface_info.scope_id}),
+             SemIR::InstId::Invalid}});
```

Comment:
This feels odd that it doesn't have an id, could we use context.constant_values().Get(interface_decl_id).inst_id()?